prog/weave-npc/main.go

```diff
@@ -312,16 +352,36 @@ func root(cmd *cobra.Command, args []string) {
 			}
 		},
 		UpdateFunc: func(old, new interface{}) {
+			defer func() {
+				if r := recover(); r != nil {
+					stopChan <- struct{}{}
+				}
+			}()
 			handleError(npc.UpdateNetworkPolicy(old, new))
 		},
 	}
 	npController = makeController(client.NetworkingV1().RESTClient(), "networkpolicies", &networkingv1.NetworkPolicy{}, npHandlers)
 
-	go nsController.Run(wait.NeverStop)
-	go podController.Run(wait.NeverStop)
-	go npController.Run(wait.NeverStop)
-
 	signals := make(chan os.Signal, 1)
+	stopChan = make(chan struct{})
+
+	go func() {
+		nsController.Run(stopChan)
+		signals <- syscall.SIGINT
```

Comment:
What's the intention here?  Looks like you are faking a signal arriving from the OS?

Reply:
I am, since that's what weave is waiting on to completely stop the application fully

Reply:
How about `os.Exit(1)` ?

Reply:
yeah, but if the weave team wants waitgroups in place, I'll likely have to use a channel of some sort, and this was already setup to use... I'm not against this, but I'm curious as to how the weave team feels.

Reply:
On line 222 we know the program has panic'd; it seems best to exit quickly and simply at that point rather than adding channels, fake signals, etc.

Reply:
I've updated this to simply `os.Exit` when one of the goroutines panics.